ligo/stablecoin/operator.ligo

```diff
@@ -0,0 +1,180 @@
+  ) : unit is block
+{ const self_transfer_policy_ : self_transfer_policy_ =
+    Layout.convert_from_right_comb ((permissions_descriptor.self : self_transfer_policy))
+; const can_transfer_from_self : bool =
+    case self_transfer_policy_ of
+      Self_transfer_permitted (u) -> True
+    | Self_transfer_denied    (u) -> False
+    end
+
+; const operator_transfer_policy_ : operator_transfer_policy_ =
+    Layout.convert_from_right_comb ((permissions_descriptor.operator : operator_transfer_policy))
+; const can_transfer_operator : bool =
+    case operator_transfer_policy_ of
+      Operator_transfer_permitted (u) -> True
+    | Operator_transfer_denied    (u) -> False
+    end
+
+; const operator : address = Tezos.sender
+
+; function owner_folding
+    ( const owner_tokens_accumulator : owner_tokens
+    ; const transfer_operation       : transfer_param
+    ) : owner_tokens is block
+  { const tokens : option (set (token_id)) =
+      Map.find_opt (transfer_operation.0, owner_tokens_accumulator)
+  ; const updated_tokens : set (token_id) =
+      case tokens of
+        None -> set [transfer_operation.1.1.0]
+      | Some (tokens) -> Set.add (transfer_operation.1.1.0, tokens)
+      end
+  } with Map.update
+      ( transfer_operation.0
+      , Some (updated_tokens)
+      , owner_tokens_accumulator
+      )
+
+; const owner_tokens : owner_tokens = List.fold
+    ( owner_folding
+    , transfer_parameters
+    , (map [] : owner_tokens)
+    )
+
+; function validate_owner
+    ( const owner  : address
+    ; const tokens : set (token_id)
+    ) : unit is block
+  { var is_operator : bool := False
+  ; if can_transfer_from_self and owner = operator
+    then { is_operator := True }
+    else if (not can_transfer_operator)
+      then failwith ("validate_operators: Operator transfer is prohibited")
+      else
+        { const operator_param : operator_param = (owner, (operator, (M_right (tokens) : operator_tokens)))
+        ; is_operator := is_operator_impl (operator_param, operators)
+        }
+  } with
+      if is_operator
+      then unit
+      else (failwith ("validate_operators: not permitted operator given") : unit)
+
+} with Map.iter
+    ( validate_owner
+    , owner_tokens
+    )
+
+(*
+ * Add operator from the given parameter and operator storage
+ *)
+function add_operator
+  ( const param     : operator_param
+  ; const operators : operators
+  ) : operators is block
+{ validate_operator_owner_is_sender (param)
+; validate_operator_token_type (param)
+; const operator_key : (owner * operator) = (param.0, param.1.0)
+} with Big_map.add (operator_key, unit, operators)
+
+(*
+ * Remove operator from the given storage
+ *)
+function remove_operator
+  ( const param     : operator_param
+  ; const operators : operators
+  ) : operators is block
+{ validate_operator_owner_is_sender (param)
+; validate_operator_token_type (param)
+; const operator_key : (owner * operator) = (param.0, param.1.0)
```

Comment:
Is it necessary to use `.1.0` to access operator? Is there no better way (access by name, not by numbers)?

Reply:
There's, but we should not neglet the fact that the conversion between comb tuples and actual records requires additional michelson instructions. Specifically in this case the type of `operator_key` is self-explanatory.

Reply:
Okay, that's quite sad :(